File: leetcode/binary-tree-level-order-traversal/binary-tree-level-order-traversal.cpp
/**************************************
* @author dujiong
* @date 2016.9.26
* @version V0.1
**************************************/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
       vector<vector<int> > ret;
       levelOrderHelp(root, ret);
       return ret;
    }
private:
    void levelOrderHelp(TreeNode *root, vector<vector<int> >& ret){
        if(root==NULL){
            return;
        }
        queue<TreeNode*> que;
  		vector<int> level;
        que.push(root);
        
        int levelNum = 1;
        
        while(!que.empty()){
            int tempNum = levelNum;
            levelNum = 0;
            for(int i=0;i<tempNum;i++){
                root = que.front();
                int temp = root->val;
                que.pop();
                level.push_back(temp);
                if(root->left){
                	que.push(root->left);
                    levelNum++;
            	}
            	if(root->right){
                	que.push(root->right);
                    levelNum++;
            	}
            }
            ret.push_back(level);
            level.clear();         
        }
    }
};

File: leetcode-golang/Binary Tree Level Order Traversal/Binary Tree Level Order Traversal.go
//二叉树的层次遍历

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	result := [][]int{}
	nodes := []*TreeNode{root}
	for len(nodes) > 0 {
		levelSlice := []int{}
		nodesCopy := nodes[:]
		nodes = []*TreeNode{}
		for _, node := range nodesCopy {
			levelSlice = append(levelSlice, node.Val)
			if node.Left != nil {
				nodes = append(nodes, node.Left)
			}
			if node.Right != nil {
				nodes = append(nodes, node.Right)
			}
		}
		result = append(result, levelSlice)
	}
	return result
}